### PR TITLE
E2E test for connection events

### PR DESCRIPTION
**Pull request by vipeller**

### CODE REVIEW

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+            //    the MQTT broker sends connection-change-notifications, and on reconnect, edgeHub can execute its
+            //    resubscription cycle.
+            //  - The way we validate that edgeHub reacts to the connection notifications is that we catch the resubscriptions
+            //    on reconnect.
+            //  - The testcode generates clients (both module and device, and direct and nested), all clients subscribe to
+            //    desired property change, so after reconnection, edgeHub should re-subscribe for all clients.
+            var (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector) = await SetupEnvironment();
+            var allClients = await ConnectClientsAsync(subscriptionChangeHandler);
+
+            await Task.Delay(2000);
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs:50]: How were these delay times chosen?

Comment (nyanzebra) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs:50]: Also, is it worthwhile making these values a constant somewhere in the test file if we are going to reuse the value?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+            }
+        }
+
+        [Fact]
+        public async Task OnReconnectionClientsGetTwinsPulled()
+        {
+            var (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector) = await SetupEnvironment();
+            var allClients = await ConnectClientsAsync(subscriptionChangeHandler);
+
+            await cloudProxyDispatcher.HandleAsync(new MqttPublishInfo("$internal/connectivity", Encoding.UTF8.GetBytes("{\"status\":\"Connected\"}")));
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs:112]: I know this is test code, but this looks similar to code above, not sure what our code guidelines are for C#, but could this go in some setup method?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs (hunk 1 above)
Comment (vipeller) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs]: those are just 'long enough" time spans, so the async processing has time to finish its job. Moved into constant

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs (hunk 2 above)
Comment (vipeller) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs]: I moved some common pieces into methods to reuse

**Review by nyanzebra — APPROVED** [review]

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+                                                    twinHandler,
+                                                    connectionHandler,
+                                                    identityProvider);
+
+            return (subscriptionChangeHandler, cloudProxyDispatcher, brokerConnector);
+        }
+
+        // This class is the one that is supposed to forward the MQTT messages to the broker.
+        // Now instead it swallows the messages, and auto-acks the RPC calls going upstream.
+        internal class NullBrokerConnector : IMqttBrokerConnector
```
Comment (vadim-kovalyov) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/ConnectionEventTest.cs:282]: seems like a good candidate to be moved to a separate file

**Review by vadim-kovalyov — APPROVED** [review]

**Pull request merged**